Handle missing account properly in stellar account exporter

satyamz opened the pull request:
### What

This will add a new metric `stellar_account_success`. Value of the `stellar_account_success` will be `1` if account exists and `0` if account doesn't exist.    


### Why
Currently, exporter returns no data if account doesn't exist and also if a particular account from the list of accounts provided via config doesn't exist then exporter breaks without reporting data regarding the accounts which exists.

Addresses: https://github.com/stellar/stellar-account-prometheus-exporter/issues/8

Signed-off-by: Satyam Zode <satyamz@users.noreply.github.com>

satyamz committed:
Handle missing account properly in stellar account exporter

Signed-off-by: Satyam Zode <satyamz@users.noreply.github.com>

jacekn commented on stellar_account_prometheus_exporter/exporter.py:
```diff
@@ -87,18 +92,18 @@ def do_GET(self):
                 if not r.ok:
-                    self.error(504, 'Error retrieving data from {}'.format(url))
-                    return
+                    m_account_exists.labels(*account_labels).set(0)
+                    continue
```
Nice and clean!
I think its worth doing the same thing for two error conditions below (missing `balances` and `subentry_count`
If there is an easy to log those error to console that would be a nice bonus but definitely don't block on it

satyamz replied:
I added metrics to denote error scenarios for the missing `balances` and `subentry_count` along with logger.

jacekn reviewed (commented):
Thank you, this is nice and clean. I added small note about handlig more errors and possible logging. Hopefully those will be easy to address

satyamz committed:
Add metrics for balance fields

Signed-off-by: Satyam Zode <satyamz@users.noreply.github.com>

jacekn commented on stellar_account_prometheus_exporter/exporter.py:
```diff
@@ -87,18 +100,24 @@ def do_GET(self):
                     self.error(504, 'Error retrieving data from {}'.format(url))
                     return
                 if not r.ok:
-                    self.error(504, 'Error retrieving data from {}'.format(url))
-                    return
+                    m_account_exists.labels(*account_labels).set(0)
+                    logger.error("Error retrieving data from {}".format(url))
+                    continue
                 if "balances" not in r.json():
-                    self.error(500, "Error - no balances found for account {}".format(account["account_id"]))
-                    return
+                    m_balance_exists.labels(*account_labels).set(0)
+                    logger.error("Error - no balances found for account {}".format(account["account_id"]))
+                    continue
                 if "subentry_count" not in r.json():
-                    self.error(500, "Error - no subentry_count found for account {}".format(account["account_id"]))
-                    return
+                    m_available_balance_exists.labels(*account_labels).set(0)
+                    logger.error("Error - no subentry_count found for account {}".format(account["account_id"]))
+                    continue
```
There may be a logic error here. If we hit the statement:
`if not r.ok:`
then we'll set one metric and then `continue`. In this case `m_balance_exists` and `m_available_balance_exists` won't be set at all.

I think it would be the simplest to just have one metric to indicate whether we got the data successfully or not for a given account. If there is _any_ problem we can set it to 0 and let operators figure it out. Data that's partially correct is effectively not correct and having one metric to indicate that is enough I think.

satyamz replied:
Initially my understanding was we wanted to provide a clear picture using granular information. Hence, I had added two additional metrics to achieve that. Idea of having single metrics sounds good too.

satyamz committed:
Use single metric to represent negative cases

This will ensure that we use a single metric to denote whether data
retrieval was successful for the given account or not.

Signed-off-by: Satyam Zode <satyamz@users.noreply.github.com>

jacekn commented on stellar_account_prometheus_exporter/exporter.py:
```diff
@@ -62,6 +65,9 @@ def do_GET(self):
         m_num_sponsoring = Gauge("stellar_account_num_sponsoring", "Stellar core account number of sponsoring entries",
                                  account_label_names, registry=self.registry)
 
+        m_account_exists = Gauge("stellar_account_success", "Displays whether or not stellar core account is present",
```
This is not entirely correct. There are error conditions, like horizon being down, that will cause errors and this metric will be 0 even if the account exists. I think it would be best to call it something generic so that it clear what it is.
```suggestion
        m_account_exists = Gauge("stellar_account_scrape_success", "Indicates whether data was gathered successfully",
```

jacekn commented on stellar_account_prometheus_exporter/exporter.py:
```diff
@@ -87,18 +95,18 @@ def do_GET(self):
+                    m_account_exists.labels(*account_labels).set(0)
+                    logger.error("Error retrieving data from {}".format(url))
+                    continue
+                if "balances" not in r.json() or "subentry_count" not in r.json():
```
Nice idea to combine it into one condition, makes code shorter :)

satyamz replied:
:)

jacekn reviewed (commented):
Thanks, it looks good. I added small suggestion about description and metric name and help text but other than that it's great!

satyamz committed:
Update stellar_account_prometheus_exporter/exporter.py

Co-authored-by: jacekn <jacekn@users.noreply.github.com>

jacekn reviewed (approved):
Very useful improvement! Thanks

satyamz commented:
Thanks for the review @jacekn!
